### PR TITLE
cat: Add the -n line numbering feature

### PR DESCRIPTION
**Pull request by PerrinJS**
Adds the -n line numbering feature to cat, with the same formatting as gnu's cat function.

### CODE REVIEW

File: Userland/Utilities/cat.cpp [hunk 1]
```diff
@@ -37,13 +41,46 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
         while (!file->is_eof()) {
             auto const buffer_span = TRY(file->read_some(buffer));
-            out("{:s}", buffer_span);
+            if (show_lines) {
+                line_count = output_buffer_with_line_numbers(line_count, buffer_span);
+            } else {
+                out("{:s}", buffer_span);
+            }
         }
+        line_count = 1;
```
Comment (LucasChollet) [inline Userland/Utilities/cat.cpp:55]: GNU's `cat` doesn't seem to reset the counter between files.

File: Userland/Utilities/cat.cpp [hunk 2]
```diff
@@ -37,13 +41,46 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
     }
 
     return files.size() != paths.size();
 }
+
+size_t output_buffer_with_line_numbers(size_t line_count, AK::Span<unsigned char> buffer_span)
```
Comment (LucasChollet) [inline Userland/Utilities/cat.cpp:61]: `Bytes` is a preferred alias for `AK::Span<unsigned char>`, here you can even use `ReadonlyBytes`

File: Userland/Utilities/cat.cpp [hunk 3]
```diff
@@ -37,13 +41,46 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
+        if ((curr_value != '\n') && (line_count != 1)) {
+            out("{:c}", curr_value);
+        } else {
+            if (line_count == 1) {
+                out("   {}\t{:c}", line_count, curr_value);
+            } else {
+                out("{:c}", curr_value);
+                if (!buffer_itter.is_end()) {
+                    // This matches the formatting of gnu's cat function
+                    out("   {:d}\t", line_count);
```
Comment (LucasChollet) [inline Userland/Utilities/cat.cpp:78]: This won't produce the right output, the number of space before the line number needs to vary.
As an example, this is an extract of `cat -n README.md` on Linux:
```
    98	* **Tom** - [tomuta](https://github.com/tomuta)
    99	* **Paul Scharnofske** - [asynts](https://github.com/asynts)
   100	* **Itamar Shenhar** - [itamar8910](https://github.com/itamar8910)
   101	* **Luke Wilde** - [Lubrsi](https://github.com/Lubrsi)
```
Off the top of my head, the correct formatting is something like:
`out("{:06}\t", line_count);`

**Review by LucasChollet — COMMENTED** [review]
Welcome to the project!

It doesn't really matter, but you can put the function in an anonymous namespace above the main to avoid having a declaration.

And also, it feels like `output_buffer_with_line_numbers` can be written in a simpler way. Feel free to try something on that side, otherwise I will give it a shot on a future review.

File: Userland/Utilities/cat.cpp (hunk 3 above)
Comment (alimpfard) [reply Userland/Utilities/cat.cpp:78]: The correct format is `"{: >6}\t"`

File: Userland/Utilities/cat.cpp (hunk 1 above)
Comment (PerrinJS) [reply Userland/Utilities/cat.cpp]: Good point, I have updated to match.

File: Userland/Utilities/cat.cpp [hunk 4]
```diff
@@ -37,13 +41,43 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
+size_t output_buffer_with_line_numbers(size_t line_count, ReadonlyBytes buffer_span)
+{
+    auto buffer_itter = buffer_span.begin();
+    while (!buffer_itter.is_end()) {
+        // to know if the next itter is the end
+        auto curr_value = *buffer_itter;
+        buffer_itter++;
+
+        if (line_count == 1 && (buffer_itter == buffer_span.begin())) {
+            out("   {: >6}\t{:c}", line_count, curr_value);
```
Comment (LucasChollet) [inline Userland/Utilities/cat.cpp:69]: `{: >6}` will take care of putting the spaces, no need to add some before
```diff
-out("   {: >6}\t{:c}", line_count, curr_value);
+out("{: >6}\t{:c}", line_count, curr_value);
```

File: Userland/Utilities/cat.cpp [hunk 5]
```diff
@@ -37,13 +41,43 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
+        } else if (curr_value != '\n') {
+            out("{:c}", curr_value);
+        } else {
+            out("{:c}", curr_value);
+            if (!buffer_itter.is_end()) {
+                // This matches the formatting of gnu's cat function
+                out("   {: >6}\t", line_count);
```
Comment (LucasChollet) [inline Userland/Utilities/cat.cpp:76]: same here

File: Userland/Utilities/cat.cpp [hunk 6]
```diff
@@ -37,13 +41,43 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
+
+size_t output_buffer_with_line_numbers(size_t line_count, ReadonlyBytes buffer_span)
+{
+    auto buffer_itter = buffer_span.begin();
+    while (!buffer_itter.is_end()) {
+        // to know if the next itter is the end
+        auto curr_value = *buffer_itter;
+        buffer_itter++;
+
+        if (line_count == 1 && (buffer_itter == buffer_span.begin())) {
```
Comment (LucasChollet) [inline Userland/Utilities/cat.cpp:68]: `buffer_itter == buffer_span.begin()`
This condition is never true as you always increment `buffer_itter` on the previous line.

Comment (LucasChollet) [reply Userland/Utilities/cat.cpp:68]: This makes me have this _wrong_ result:
```
$cat -n README.md 
# SerenityOS and Ladybird
        1	
        2	SerenityOS is a graphical Unix-like operating system for x86-64 computers.
...
```

File: Userland/Utilities/cat.cpp [hunk 7]
```diff
@@ -37,13 +41,43 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
+            out("   {: >6}\t{:c}", line_count, curr_value);
+        } else if (curr_value != '\n') {
+            out("{:c}", curr_value);
+        } else {
+            out("{:c}", curr_value);
+            if (!buffer_itter.is_end()) {
```
Comment (LucasChollet) [inline Userland/Utilities/cat.cpp:74]: One issue that I can see here is that you're testing for the end of a relatively small buffer, not the end of the file. If the 32k buffer ends by `\n`, you will be missing a line count.

To fix that, I would take another approach:
Let's declare a `struct` that store the state, with the line count and a `bool` `display_line_number`.
The struct is passed to the function and also returned from it (as you did with `line_count`).
Then `output_buffer_with_line_numbers` can simply be:
```cpp
for (auto const c : buffer) {
	if (display_line_number) {
		out("{: >6}\t", line_count);
		display_line_number = false;
	}
	if (c == '\n')
		display_line_number = true;
	out("{:c}", c);
}
```

Comment (PerrinJS) [reply Userland/Utilities/cat.cpp:74]: Yea, sorry about the bad code I was rushing as I haven't had much time this week to work on projects. Will have another go at it on Sunday. Thanks for the help though.

File: Userland/Utilities/cat.cpp [hunk 8]
```diff
@@ -37,11 +58,19 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
 
     TRY(Core::System::pledge("stdio"));
 
+    LineTracker line_tracker;
+    line_tracker.line_count = 1;
+    line_tracker.display_line_number = true;
```
Comment (LucasChollet) [inline Userland/Utilities/cat.cpp:63]: nit: Let's default-initialize that

**Review by LucasChollet — APPROVED** [review]
One small nitpick, otherwise LGTM!

File: Userland/Utilities/cat.cpp [hunk 9]
```diff
@@ -11,15 +11,36 @@
 #include <LibCore/System.h>
 #include <LibMain/Main.h>
 
+struct LineTracker {
+    size_t line_count;
+    bool display_line_number;
```
Comment (nico) [inline Userland/Utilities/cat.cpp:16]: https://github.com/SerenityOS/serenity/pull/24382/files#r1648267581 meant "please add default initializers here", i.e.

```
struct LineTracker {
    size_t line_count { 0 };
    bool display_line_number { false };
};
```

At least that's how I understood that comment. All struct fields should have initializers :)

Comment (PerrinJS) [reply Userland/Utilities/cat.cpp:16]: Ah, Thanks. I wasn't sure, not super good with C++ yet.
I'll sort it out tomorrow.

Comment (PerrinJS) [reply Userland/Utilities/cat.cpp:16]: I see what happened. I forgot to actually append the changes. Sorry for the confusion.

**Comment by PerrinJS** [comment]
Please disregard the changes prior to this as it still does not contain the updates. Thank you for bearing with me.

**Review by LucasChollet — APPROVED** [review]

File: Userland/Utilities/cat.cpp [hunk 10]
```diff
@@ -11,15 +11,36 @@
 #include <LibCore/System.h>
 #include <LibMain/Main.h>
 
+struct LineTracker {
+    size_t line_count = 1;
```
Comment (nico) [inline Userland/Utilities/cat.cpp:15]: fyi: `size_t line_count { 1 };` is what we use in most other places. I'll merge this, but if you end up touching this file again, maybe you could fix it up then.

**Comment by nico** [comment]
Thanks!

**Pull request merged**